src/stamps.ts

```diff
@@ -1,28 +1,235 @@
+ * @param minTTL Minimal TTL of the stamp to be usable by the proxy
+ *
+ * @returns Filtered stamps soltered by usage
+ */
+export function filterUsableStamps(
+  stamps: DebugPostageBatch[],
+  depth: number,
+  amount: string,
+  maxUsage: number,
+  minTTL: number,
+): DebugPostageBatch[] {
+  const usableStamps = stamps
+    // filter to get stamps that have the right depth, amount and are not fully used or expired
+    .filter(s => s.usable && s.depth === depth && s.amount === amount && getUsage(s) < maxUsage && s.batchTTL > minTTL)
+    // sort the stamps by usage
+    .sort((a, b) => (getUsage(a) < getUsage(b) ? 1 : -1))
+
+  // return the all usable stamp sorted by usage
+  return usableStamps
+}
+
+/**
+ * Buy new postage stamp and wait until it is usable
+ *
+ * @param depth Postage stamps depth
+ * @param amount Postage stamps amount
+ * @param beeDebug Connection to debug endpoint for checking/buying stamps
+ * @param pollingFrequency (optional) How often should the stamp be checked in ms, default 1000
+ *
+ * @returns Newly bought postage stamp
  */
-export function shouldReplaceStamp({ POSTAGE_STAMP }: PostageStamps = {}): boolean {
-  return Boolean(POSTAGE_STAMP)
+export async function buyNewStamp(
+  depth: number,
+  amount: string,
+  beeDebug: BeeDebug,
+  pollingFrequency = 1_000,
+): Promise<DebugPostageBatch> {
+  return new Promise(async (resolve, reject) =>
+    beeDebug
+      .createPostageBatch(amount, depth)
+      .then(async (batchId: BatchId) => {
+        waitUntilStampUsable(batchId, beeDebug, pollingFrequency)
+          .then(resolve)
+          .catch(e => {
+            logger.error('error: failed to wait until postage stamp is usable', e)
+            reject(e)
+          })
+      })
+      .catch(e => {
+        logger.error('error: failed to buy postage stamp', e)
+        reject(e)
+      }),
+  )
+}
+
+export class StampsManager {
+  private stamp?: string
+  private usableStamps?: DebugPostageBatch[]
+  public readonly enabled: boolean
+  private interval?: ReturnType<typeof setInterval>
+
+  constructor(
+    {
+      POSTAGE_STAMP,
+      POSTAGE_DEPTH,
+      POSTAGE_AMOUNT,
+      POSTAGE_USAGE_THRESHOLD,
+      POSTAGE_USAGE_MAX,
+      POSTAGE_TTL_MIN,
+      POSTAGE_REFRESH_PERIOD,
+    }: Args = {},
+    BEE_DEBUG_API_URL?: string,
+  ) {
+    this.enabled = true
+
+    if (POSTAGE_STAMP) this.stamp = POSTAGE_STAMP
+    else if (POSTAGE_DEPTH && POSTAGE_AMOUNT && BEE_DEBUG_API_URL) {
+      const refreshPeriod = Number(POSTAGE_REFRESH_PERIOD || 60_000)
+      this.start(
+        Number(POSTAGE_DEPTH),
+        POSTAGE_AMOUNT,
+        Number(POSTAGE_USAGE_THRESHOLD || '0.7'),
+        Number(POSTAGE_USAGE_MAX || '0.95'),
+        Number(POSTAGE_TTL_MIN || (refreshPeriod / 1_000) * 5),
+        refreshPeriod,
+        new BeeDebug(BEE_DEBUG_API_URL),
+      )
+    } else if (POSTAGE_DEPTH || POSTAGE_AMOUNT || BEE_DEBUG_API_URL) {
```

Comment:
```suggestion
    } else if (!POSTAGE_DEPTH || !POSTAGE_AMOUNT || !BEE_DEBUG_API_URL) {
```

Reply:
This is not correct, the idea is that you can start the gateway without providing hardcoded postage stamp with `POSTAGE_STAMP` and with the autobuy disabled (meaning there is not `POSTAGE_DEPTH`, `POSTAGE_AMOUNT` and `BEE_DEBUG_API_URL` provided). I agree this is not obvious and could be improved.

So to summarise, there are 3 modes of operation:
- it just proxies without replacing postage stamp
- it replaces with hardcoded postage stamp
- it replaces with some stamp that is automatically bought with the autobuy feature